backend/accounts/transaction.go

```diff
@@ -133,5 +142,122 @@ type OrderedTransactions []*TransactionData
 // The input list is modified in place and must not be used anymore after calling this function.
 func NewOrderedTransactions(txs []*TransactionData) OrderedTransactions {
 	sort.Sort(sort.Reverse(byHeight(txs)))
+
+	balance := big.NewInt(0)
+	for i := len(txs) - 1; i >= 0; i-- {
+		tx := txs[i]
+		switch tx.Type {
+		case TxTypeReceive:
+			balance.Add(balance, tx.Amount.BigInt())
+		case TxTypeSend:
+			balance.Sub(balance, tx.Amount.BigInt())
+			// Subtract fee as well.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		case TxTypeSendSelf:
+			// Subtract only fee.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		}
+		tx.Balance = coin.NewAmount(balance)
+	}
 	return txs
 }
+
+// TimeseriesEntry contains the balance of the account at the given time.
+type TimeseriesEntry struct {
+	Time  time.Time
+	Value coin.Amount
+}
+
+// MarshalJSON serializes the entry as JSON.
+func (c *TimeseriesEntry) MarshalJSON() ([]byte, error) {
+	int64, err := c.Value.Int64()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(map[string]interface{}{
+		"time":  c.Time.Unix(),
+		"value": int64,
+	})
+}
+
+// findLastIndexBeforeTime returns the index of the last transaction with a time before `target`,
+// starting at `start`.
+// Unconfirmed txs are ignored.
+// If there is no such transaction before `target`, nil is returned.
+// Note: the transactions are ordered from newest to oldest, so the returned index will be smaller
+// or equal to `start` (advancing backwards).
+func (txs OrderedTransactions) findLastIndexBeforeTime(
+	start int, target time.Time) (*int, error) {
+	currentIndex := start
+	for {
+		tx := txs[currentIndex]
+
+		if tx.Height == 0 {
+			// We reached the unconfirmed txs. We ignore those. The current index is the last tx
+			// before the target time.
+			return &currentIndex, nil
+		}
+		txTime := tx.Timestamp
+		if txTime == nil {
+			return nil, errp.WithStack(errors.ErrNotAvailable)
+		}
+		if target.Before(*txTime) {
+			if currentIndex == start {
+				// No tx before target, return nil.
+				return nil, nil
+			}
+			// We reached the first tx beyond target. Return the previous index, which is the last
+			// tx before target.
+			currentIndex++
+			return &currentIndex, nil
+		}
+		if currentIndex == 0 {
+			return &currentIndex, nil
+		}
+		currentIndex--
```

Comment:
Could use binary search to reduce this loop from O(N) to O(log N) and the loop in `Timeseries` from `O(M*N)` to `O(M*log N)` by extension for those who have many transactions.
https://golang.org/pkg/sort/#Search

Reply:
It is already O(N) (n=len(txs)). The index strictly advances through the slice.

Reply:
so, you're saying O(N) is better than O(log N)?

Reply:
Ah, haha brain was off it seems ;) 

You are right, but I would skip it for simpler code. Binary search here will only help if you have lots of transactions within one interval, but then also probably only measurable if it is quite a large number.

Reply:
But try it! I think it might actually simplify code. You need to provide only the lookup-at-index function:

```go
func Search(n int, f func(int) bool) int
```

Reply:
It was designed for cases exactly like this one:

> Search uses binary search to find and return the smallest index i in [0, n) at which f(i) is true

Reply:
I added it in a new commit, please check. Nice idea!